src/Symfony/Component/AutoMapper/AutoMapper.php

```diff
@@ -0,0 +1,122 @@
+<?php
+
+namespace Symfony\Component\AutoMapper;
+
+use Psr\Container\ContainerInterface;
+use Symfony\Component\AutoMapper\Exception\RuntimeException;
+use Symfony\Component\AutoMapper\Attributes\MapIf;
+use Symfony\Component\AutoMapper\Attributes\MapTo;
+use Symfony\Component\AutoMapper\Attributes\MapWith;
+use Symfony\Component\PropertyAccess\PropertyAccessorInterface;
+
+/**
+ * Object to object mapper.
+ *
+ * @experimental
+ * @author Antoine Bluchet <soyuka@gmail.com>
+ */
+final class AutoMapper implements AutoMapperInterface
+{
+    public function __construct(private ?PropertyAccessorInterface $propertyAccessor = null, private ?ContainerInterface $serviceLocator = null)
+    {
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function map(object $object, object|string|null $to = null): mixed
+    {
+        $refl = new \ReflectionClass($object);
+
+        if (!$to) {
+            $to = $this->getAttribute($refl, MapTo::class, true)->to;
+        }
+
+        $arguments = [];
+        if (is_object($to)) {
+            $toRefl = new \ReflectionClass(get_class($to));
+            $mapped = $to;
+        } else {
+            $toRefl = new \ReflectionClass($to);
+            $mapped = $toRefl->newInstanceWithoutConstructor();
+        }
+
+        $constructor = $toRefl->getConstructor();
+        foreach ($constructor?->getParameters() ?? [] as $parameter) {
+            $arguments[$parameter->getName()] = $parameter->isDefaultValueAvailable() ? $parameter->getDefaultValue() : null;
+        }
+
+        foreach ($refl->getProperties() as $property) {
+            $propertyName = $property->getName();
+            $mapTo = $this->getAttribute($property, MapTo::class)?->to ?? $propertyName;
+            if (!$toRefl->hasProperty($mapTo)) {
+                continue;
+            }
+
+            $value = $this->propertyAccessor ? $this->propertyAccessor->getValue($object, $propertyName) : $object->{$propertyName};
+            $mapIf = $this->getCallable($this->getAttribute($property, MapIf::class)?->if);
+            if (is_callable($mapIf) && false === $this->call($mapIf, $value, $object)) {
+                continue;
+            }
+
+            $mapWith = $this->getCallable($this->getAttribute($property, MapWith::class)?->with);
+            if (is_callable($mapWith)) {
+                $value = $this->call($mapWith, $value, $object);
+            }
+
+            if (is_object($value) && $to = $this->getAttribute(new \ReflectionClass(get_class($value)), MapTo::class)?->to) {
+                $value = $this->map($value, $to);
```

Comment:
We'll need to think about a "depth" here to avoid circular references with relationships

Reply:
I'd suggest throwing if it looks like we've a circular dependency instead of managing max-depth.

Reply:
I commonly map both sides of a relation, so (in entities) I have `Product.category` and `Category.products`. So going from Product -> ProductDto would trigger that circular problem. But... HOW to solve it, I don't know. In micromapper, I have a "depth" where, if the depth is hit, the last level is instantiated but not populated, but I'm not sure if it makes sense. So, for example, if max depth = 1, then when going to Product -> ProductDTO, you end up with:

```
ProductDto
   id: 5
   category: CategoryDto
        id: 10
        product: ProductDto
            id: null (this is beyond the depth, it instantiates the object, but doesn't populate it)
```

Not sure if that's great - the circular relations are tough :/

Reply:
Yeah I know how to fix that I'll let you know when it's implemented.

Reply:
`SplObjectStorage<from, to>` looks like the perfect candidate here